include/deal.II/fe/fe_poly_tensor.h

```diff
@@ -224,6 +224,122 @@ class FE_PolyTensor : public FiniteElement<dim, spacedim>
+   * three flags.
+   *
+   * There is also a pattern for the sign change of the permuted shape functions
+   * that depends on the combination of the switches. In the above example it
+   * would be
+   *
+   * @verbatim
+   * ___________
+   * |           |
+   * |  +  +  -  |
+   * |           |
+   * |  +  +  -  |
+   * |           |
+   * |  +  +  -  |
+   * |___________|
+   * @endverbatim
+   *
+   * @note This function is not a pure virtual function but still must be implemented
+   * in finite elements classes derived from FE_PolyTensor since the permutation
+   * pattern and the pattern of sign changes depends on how the finite element
+   * distributes the local dofs on the faces. Each (vector) finite element class
+   * that does not have a robust implementation in Deal.II throws an exception
+   * if there is a cell in the triangulation that has non-standard, flipped or
+   * rotated faces.
+   */
+  virtual void
+  initialize_quad_dof_index_permutation_and_sign_change();
+
+  /**
+   * For faces with non-standard face_orientation in 3D, the dofs on faces
+   * (quads) have to be permuted in order to be combined with the correct
+   * shape functions and additionally can change the sign. Given a local dof @p index on a quad, return the
+   * sign of the permuted shape function, if the face has non-standard
+   * face_orientation, face_flip or face_rotation. In 2D and 1D there is no need
+   * for permutation and consequently it does nothing in this case.
+   *
+   * The permutation itself is returned by
+   * adjust_quad_dof_index_for_face_orientation implemented in the interface
+   * class FiniteElement<dim>.
+   */
+  bool
+  adjust_quad_dof_sign_for_face_orientation(const unsigned int index,
+                                            const unsigned int face_no,
+                                            const bool         face_orientation,
+                                            const bool         face_flip,
+                                            const bool face_rotation) const;
+
+  /**
+   * For faces with non-standard face_orientation in 3D, the dofs on faces
+   * (quads) have to be permuted in order to be combined with the correct
+   * shape functions. For elements with shape that have an orientation relative
+   * to faces finding the permutations is not enough. We possibly need to change
+   * the sign of the shape function on faces depending on the combination of the
+   * three bool flags face_orientation, face_flip and face_rotation.
+   *
+   * The constructor of this class fills this table with ones, i.e.,
+   * no sign change at all. Derived finite element classes have to
+   * fill this Table with the correct values, see the documentation in
+   * GeometryInfo<dim> and
+   * this @ref GlossFaceOrientation "glossary entry on face orientation".
+   */
+  std::vector<Table<2, bool>> adjust_quad_dof_sign_for_face_orientation_table;
```

Comment:
This may be true. I implemented it in the same terms as the table <code>adjust_quad_dof_index_for_face_orientation_table</code> since this way one does not need to rethink the structure of the implementation for <code>adjust_quad_dof_sign_for_face_orientation_table</code> too much. This may also help other people to fix other elements by having a look at the implementation of `RaviartThomas<dim>`. Besides, I understand your concern but is the overhead really significant here? In most codes the finite element is constructed once at the beginning and the code is not called repeatedly. If you agree I would like to leave it as is.

Reply:
If you think this is the right way to do it, then let's leave it!